OpenRA.Game/Network/Connection.cs

```diff
@@ -139,34 +184,74 @@ public void Dispose()
 	sealed class NetworkConnection : EchoConnection
 	{
 		readonly TcpClient tcp;
+		readonly IPEndPoint endpoint;
 		readonly List<byte[]> queuedSyncPackets = new List<byte[]>();
 		volatile ConnectionState connectionState = ConnectionState.Connecting;
 		volatile int clientId;
 		bool disposed;
 
-		public NetworkConnection(string host, int port)
+		public override IPEndPoint EndPoint { get { return endpoint; } }
+
+		public NetworkConnection(ConnectionTarget target)
 		{
-			try
+			var queue = new BlockingCollection<TcpClient>();
+
+			foreach (var endpoint in target.GetConnectEndPoints())
 			{
-				tcp = new TcpClient(host, port) { NoDelay = true };
-				new Thread(NetworkConnectionReceive)
+				new Thread(() =>
+				{
+					try
+					{
+						var client = new TcpClient(endpoint.AddressFamily) { NoDelay = true };
+						client.Connect(endpoint.Address, endpoint.Port);
+
+						try
+						{
+							queue.Add(client);
+						}
+						catch (InvalidOperationException)
+						{
+							// Another connection was faster, close this one.
+							client.Close();
+						}
+					}
+					catch (Exception ex)
+					{
+						Log.Write("client", "Failed to connect to {0}: {1}", endpoint, ex.Message);
+					}
+				})
 				{
-					Name = GetType().Name + " " + host + ":" + port,
+					Name = "{0} (connect to {1})".F(GetType().Name, endpoint),
 					IsBackground = true
-				}.Start(tcp.GetStream());
+				}.Start();
 			}
-			catch
+
+			// Wait up to 5s for a successful connection. This should hopefully be enough because such high latency makes the game unplayable anyway.
+			if (queue.TryTake(out tcp, 5000))
```

Comment:
This also fails rather awkwardly if something else is already using the port, blocking for the full 5 seconds while displaying no addresses:

<img width="486" alt="Screenshot 2020-02-08 at 00 21 11" src="https://user-images.githubusercontent.com/167819/74075170-15da2380-4a09-11ea-8194-17a6a275502c.png">

The simple repro case is to start a dedicated server and then try to create a server ingame on the same port.

Reply:
The second issue has a different cause. I checked the wrong list in `Server.cs` so you could actually end up with a server instance without any socket.

Reply:
The connecting process is now done in its own thread with error handling slightly reworked to show a more specififc error if name resolution fails instead of just the generic "Connection Lost".